c/shared/source/uforead/uforead.c

```diff
@@ -1813,10 +1903,70 @@ static bool parseFontInfoFDArray(ufoCtx h, xmlNodePtr cur) {
     return true;
 }
 
+/* set each GLIF_REC's iFD based on the FDArraySelect groups*/
+static bool parseFDArraySelectGroup(ufoCtx h, char* FDArraySelectKeyName, xmlNodePtr cur) {
+    GLIF_Rec *g;
+
+    if (!xmlStrEqual((cur)->name, (const xmlChar *) "array"))
+        return false;
+
+    char* FDIndex = memNew(h, sizeof(char) * 2);
+    strncpy(FDIndex, FDArraySelectKeyName + 14, 2);
+    if (FDIndex[1] == '.')
+        FDIndex[1] = '\n';
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        char* glyphName = parseXMLKeyValue(h, cur);
+        g = findGLIFRecByName(h, glyphName);
+        int FDIndexInt = atoi(FDIndex);
+        if (g != NULL) {
+            if (FDIndexInt >= h->top.FDArray.cnt - 1) {  /* one cnt reserved for default FDArray at this point */
+//                message(h, "Glyph %s is assigned to font dict index %d which is not defined. Will be assigned to default font dict at index 0 instead.\n", glyphName, FDIndexInt);
+                FDIndexInt  = 0;
+            }
+            g->iFD = FDIndexInt;
+        }
+        cur = cur->next;
+    }
+    memFree(h, FDIndex);
+    return true;
+}
+
+static bool parseCIDMap(ufoCtx h, xmlNodePtr cur) {
+    h->parseState.CIDMap = true;
+    h->top.sup.flags |= ABF_CID_FONT;
```

Comment:
I'm not sure we should be setting ABF_CID_FONT here because you can delete the other required keys and if `com.adobe.type.postscriptCIDMap` still exists the UFO will be read as CID-keyed and then fail because not everything required is set. I think it might make sense to have a preparser to check for all required keys first. We only need to read top level keys and look for name, ROS, CIDMap, and FDArrays. If any of those are missing we should bail on trying to read as CID-keyed. If we do that then this line is fine because we can set the flag if CIDMap comes before name and ROS.

Reply:
My latest commit implements these checks, including FDArray. If we'd like to remove some keys from being required, I can just comment those out later.

Do we want to error/warn the user if there are missing required keys but the font looks like it should be CID-keyed? (like if one or more CID-keyed keys are present, but not all of them). I'll add tests for this once we make a decision on this.